Support serializable Lambdas as nested inputs

alllex opened the pull request:
Serializable lambdas are now supported as nested inputs for a task.

Additionally, the problem with absolute path dependent script names is resolved, fixing the up-to-date checks in a location-independent manner.

Fixes #11703

Related to #21085

alllex commented:
@bot-gradle test qfl

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"36928fba8630e00d5812b8cf4f5af51dc43b5acc","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [QuickFeedbackLinux build](https://builds.gradle.org/viewQueued.html?itemId=53265418)

alllex commented:
@bot-gradle test qfl

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"38db87cce1602a9b2e84886d75af8a15c1b7cf11","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [QuickFeedbackLinux build](https://builds.gradle.org/viewQueued.html?itemId=53265604)

alllex commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e143fc6ba456d06cdfd2cdf9bb3533437cef6694","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=53265798)

alllex commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"e33927d3dc05e86a23dbc7004567874468518eea","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=53266859)

wolfs reviewed (commented):
Looks good. I left some small comments.

alllex commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy:
```diff
@@ -97,41 +103,72 @@ class LambdaInputsIntegrationTest extends AbstractIntegrationSpec implements Val
+        when:
+        run 'myTask', '-Pchanged'
+        then:
+        executedAndNotSkipped(':myTask')
```
I think a word is missing from you comment and I can't infer what you meant.
What should we do with the message?

wolfs replied:
Right, I clarified the message.

alllex commented:
@bot-gradle test qfl

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"32695006874647d59f38f7d68e700857e2b6463a","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [QuickFeedbackLinux build](https://builds.gradle.org/viewQueued.html?itemId=53345412)

alllex commented:
@wolfs I applied all your suggestions, and also added a new test that checks tracking for a non-Action serializable lambda.
The consumer-based task was nicely reused, because I added a `SerializableConsumer` interface as we did in some other tests.

alllex commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"32695006874647d59f38f7d68e700857e2b6463a","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=53346625)

wolfs commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy:
```diff
@@ -75,19 +76,24 @@ class LambdaInputsIntegrationTest extends AbstractIntegrationSpec implements Val
         """
     }
 
+    @ToBeFixedForConfigurationCache(because = "Non-serializable lambdas are not supported in configuration cache")
```
I saw you added a link to the issue about better reporting for non-serializable lambdas. I am not sure this is necessary, given that this is not a test which would check the error message when configuration cache is on. I was referring to an issue about being able to store non-serializable lambdas in the configuration cache. In that case, this issue would go into the `because = ""` of the `ToBeFixedForConfigurationCache` annotation.

The `ToBeImplemented` annotation is more for tests which would check a future behavior, and this test doesn't seem to be that. You'd use `@Issue` here if the test checks a behavior in some issue. Though that is also not the case here.

All in all, I'd only link to an issue about "Support non-serializable lambdas with configuration caching" and then use a link to this issue in `because`.

alllex replied:
Would this issue work for that purpose?
https://github.com/gradle/gradle/issues/21109

wolfs replied:
Yes, that is perfect.

alllex replied:
Done

wolfs commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy:
```diff
@@ -373,16 +492,35 @@ class LambdaInputsIntegrationTest extends AbstractIntegrationSpec implements Val
+    /**
+     * Configuration cache infrastructure makes sure that {@code Action} instances are always serializable.
+     */
+    private static String serializableActionLambdaWritingFile(String constantName, String outputString) {
```
```suggestion
    private static String actionLambdaWritingFile(String constantName, String outputString) {
```

Maybe we should drop the `serializable` from the method name, given that `Action` is not serializable from the start. With the comment I think it would make sense.

alllex replied:
Done

wolfs reviewed (commented):
I left some more comments.

wolfs reviewed (approved):
LGTM!

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=53359160) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildId=53359160) failed.

alllex commented:
@bot-gradle test and merge

bot-gradle commented:
Your PR is queued. See [the queue page](https://gradle-bot.grdev.net/queue.html) for details.

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=53369951) for you.
